pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package couchdbstore
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"sync"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+	dbs           map[string]*CouchDBStore
+	mux           sync.RWMutex
+}
+
+const (
+	blankHostErrMsg           = "hostURL for new CouchDB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+)
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client, dbs: map[string]*CouchDBStore{}}, nil
+}
+
+// CreateStore creates a new store with the given name.
+func (p *Provider) CreateStore(name string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	err := p.couchDBClient.CreateDB(context.Background(), name)
+
+	return err
+}
+
+// OpenStore opens an existing store with the given name and returns it.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	exists, err := p.storeExists(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, storage.ErrStoreNotFound
+	}
+
+	// Check cache first
+	store, exists := p.dbs[name]
```

Comment:
Seems a bit awkward that you check if the store exists using storeExists and then you check again if it exists here. Maybe it would be better to get rid of storeExists, unless it's being used elsewhere?

Reply:
Done.